Custom Robotest Image

wadells opened the pull request:
## Description
This PR introduces a new cluster image `robotest.tar` for our automated integration testing.  The contents of the image are minimal, this largely the build & infra work to support an the new artifact.

Why? Gravity, like the kubernetes it wraps has a plethora of customization options.  Running it without any of these customization options set and without an application running inside the cluster is not realistic for our customers. We want to move towards more realistic integration scenarios.

See https://github.com/gravitational/gravity/issues/1908 for more context.

## Type of change
* New feature (non-breaking change which adds functionality)
* Internal change (not necessarily a bug fix or a new feature)

## Linked tickets and other PRs
* Closes https://github.com/gravitational/gravity/issues/1908
* Related to https://github.com/gravitational/robotest/issues/160 and https://github.com/gravitational/robotest/issues/197

## TODOs
- [x] Self-review the change
- [x] Perform manual testing
- [ ] Write documentation
- [x] Address review feedback

## Implementation
I've kept the implementation minimal for the initial version, preferring an incremental approach to trying for a more substantial image, as a more waterfall-y design first strategy wasn't providing timely results.

Follow up work that is explicitly out of scope *for this PR* includes:
* Integrating with upgrade testing https://github.com/gravitational/gravity/issues/1915
* Implementing a NFS provisioner https://github.com/gravitational/gravity/issues/1909
* Allowing the contents to be tweaked in an ad-hoc manner https://github.com/gravitational/gravity/issues/1923
* 

## Testing done
TBD

wadells committed:
Remove old assets/robotest content.

This is unused, and I want to use the namespace assets/robotest for a
cluster image.

wadells committed:
Remove unused robotest targets.

These have not been used to produce robotest artifacts since the repo
was open sourced.

robotest-installer-ready was removed in df4bcc75.

robotest-installer stopped being called 383813b5.

robotest-publish-gravity was only called by robotest-installer.

wadells committed:
Add custom Robotest application image.

wadells committed:
Remove trailing / from top directory.

This is not needed, and results in harmless but unsightly `//` in the
middle of paths in many cases.

wadells committed:
Remove unused Makefile variables.

LOCAL_ETCD_DIR, GRAVITY_ASSETS_DIR and TELEPORT_DIR haven't been
used since 213fb448c66edd55cd476e079849eaea42c300a9.

wadells committed:
Replace $(GRAVITY_BUILDDIR)/telekube.tar with $(TELEKUBE_OUT).

We did this inconsistently in different places in the makefile.  Using
the same variable means these values change together.

wadells committed:
Move gravity --state-dir cache to build dir.

Previously, gravity builds used /var/lib/gravity/opscenter/read as
a --state-dir. Now this cache lives at $(GRAVITY_BUILDDIR)/state.

Advantages:

 - `make production telekube` no longer needs to write to /var/lib/gravity
   (requiring modified permissions for this directory or elevated
   privileges for the build process). This makes building gravity more
   friendly first time users & new developers.

Disadvantages:

 - If packages were shared between runs, they're now removed by `make clean`
   and no longer shared between gravity versions with different build
   numbers.

Integration considerations:

 - /var/lib/gravity/opscenter/read is 300GB on jenkins.  It is on
   a dedicated disk.  Once this is integrated for long enough,
   we can remove this.
 - This is consistent with what @knisbet implemented for mage in #1917.

r0mant commented on Makefile:
```diff
@@ -522,15 +520,35 @@ scan-artifacts: telekube
+		-o $(TELEKUBE_OUT)
+
+$(GRAVITY_OUT):
+	$(MAKE) -C build.assets build
```
Nit: This will build all binaries, not just `gravity`.

wadells replied:
What is the correct way to build just `GRAVITY_OUT`?  I am having a deal of trouble figuring that out, because the binary targets are lumped together under a parallel dockerized make in `build-in-container` -- which I didn't really want to take as a pretty slow (2m22s on my system) `.PHONY` dependency.

This `$(GRAVITY_OUT)` target allows me to cheat and make sure a binary exists on the FS without taking the costly `.PHONY` dependency.  It is still prone to failure because it in turn doesn't depend on any of the upstreams that feed into `$(GRAVITY_OUT), and thus won't get rebuilt if a source changes -- but that is rampant in our makefiles and not a problem I want to tackle in this PR.

r0mant replied:
Hmm, IIRC there's a `BINARIES` variable that sets the binaries to build so you could try to use something like

```
BINARIES=gravity $(MAKE) -C build.assets build
```

wadells replied:
Fixed in beeb7c5.  This may lead to non-parallel rebuilds in some cases.

r0mant commented on Makefile:
```diff
@@ -652,63 +670,17 @@ wizard-gen:
-robotest-installer: INSTALLER_FILE := $(subst /,-,$(subst :,-,$(ROBOTEST_APP_PACKAGE)))-$(GRAVITY_VERSION)-installer.tar.gz
-robotest-installer: INSTALLER_URL := $(ROBO_BUCKET_URL)/$(INSTALLER_FILE)
-robotest-installer: robotest-publish-gravity
-robotest-installer:
```
Is this no longer used? How will robotest use the new application for install scenarios? IIRC currently it builds it locally?

wadells replied:
This is unused for many years.  More context in the commit message:

https://github.com/gravitational/gravity/pull/1967/commits/dd357c87cab1303eb537ea4c175546df2995cf56

The `robotest-image` target produces the new installer tarball which will be used for all scenarios.  I don't have a strong opinion on the name.

r0mant commented on assets/robotest/resources/app.yaml:
```diff
@@ -0,0 +1,163 @@
+apiVersion: bundle.gravitational.io/v2
```
"bundle" is the legacy version/kind, let's use new header:

```yaml
apiVersion: cluster.gravitational.io/v2
kind: Cluster
```

wadells replied:
Fixed in 9bc4135.

r0mant commented on assets/robotest/resources/app.yaml:
```diff
@@ -0,0 +1,163 @@
+  status:
+    job: file://status.yaml
+  postInstall:
+    job: file://status.yaml
```
Is "status.yaml" used as a placeholder for other hooks for now?

r0mant commented on assets/robotest/resources/app.yaml:
```diff
@@ -0,0 +1,163 @@
+  openebs:
+    enabled: true
+systemOptions:
+  runtime:
```
systemOptions.runtime is also an obsolete field, I'd just remove it.

And systemOptions.docker as well, overlay2 is the default driver.

So I'd remove the entire systemOptions section for now.

wadells replied:
Removed in 79d539d.

r0mant commented on assets/robotest/resources/app.yaml:
```diff
@@ -0,0 +1,163 @@
+        - path: /var/lib/gravity
+          capacity: "10GB"
+          filesystems: ["xfs", "ext4"]
+      volumes:
```
I would add the same custom volume to other node profiles as well.

wadells replied:
Added in 82fef03.

r0mant commented on assets/robotest/resources/install.yaml:
```diff
@@ -0,0 +1,22 @@
+         - name: echoserver-install
+           image: quay.io/gravitational/debian-tall:buster
+           command:
+             - /usr/local/bin/helm
```
I would also set a release name (`--name xxx`) so it can be used in "helm upgrade" command in the upgrade hook.

wadells committed:
Use cluster instead of bundle in robotest manifest.

Bundle is legacy and was replaced by cluster, per:

  https://github.com/gravitational/gravity/pull/1967#discussion_r465924415

wadells committed:
Remove systemOptions from robotest manifest.

All of this is out of scope for current testing per:

  https://github.com/gravitational/gravity/pull/1967#discussion_r465925381

wadells committed:
Add custom mount to Auto & Master node profiles.

wadells committed:
Split upgrade testing artifacts.

Bump robotest in order to accommodate per-test gravity installers.  Then
continue to use telekube.tar for upgrade tests, while moving to
robotest.tar for all other testing.

r0mant commented on Jenkinsfile:
```diff
@@ -84,6 +84,7 @@ timestamps {
         robotest : {
           if (env.RUN_ROBOTEST == 'run') {
             withCredentials([
+                [$class: 'StringBinding', credentialsId: 'GET_GRAVITATIONAL_IO_APIKEY', variable: 'GET_GRAVITATIONAL_IO_APIKEY'],
```
Still need this with https://github.com/gravitational/gravity/pull/2023?

wadells replied:
This is no longer needed because of #1997. #2023 isn't relevant to this PR because it targets 7.0.x and upgrade-via which hasn't hit master yet (as far as I know).

There will be a bit of merge headache between https://github.com/gravitational/gravity/issues/1915 and the upgrade via work, but nothing out of the ordinary given our port strategy.
